spark-cassandra-connector/src/main/scala/com/datastax/spark/connector/rdd/reader/GettableDataToMappedTypeConverter.scala

```diff
@@ -0,0 +1,254 @@
+    * and for everything else uses
+    * [[com.datastax.spark.connector.mapper.DefaultColumnMapper DefaultColumnMapper]] */
+  private def columnMapper[U : TypeTag]: ColumnMapper[U] = {
+    val tpe = TypeTag.synchronized(typeTag[U].tpe)
+    if (tpe.typeSymbol.fullName startsWith "scala.Tuple")
+      new TupleColumnMapper[U]
+    else if (isJavaBean) 
+      new JavaBeanColumnMapper[U]()(ReflectionUtil.classTag[U])
+    else
+      new DefaultColumnMapper[U]
+  }
+
+  /** Returns a converter that can convert a Cassandra value of given `ColumnType` to Scala type `U`. */
+  private def converter[U : TypeTag](columnType: ColumnType[_]): TypeConverter[U] = {
+    /*
+    This method does not immediately call `TypeConverter.forType`, because `forType` doesn't know
+    the Cassandra `ColumnType` and therefore it can't instantiate appropriate type converters for
+    Cassandra user-defined-types, which could be nested. Therefore, for types that can nest other types
+    i.e. collections, options or UDTs, we make sure to use this method for any needed nested converters.
+    E.g. for a list type, we recursively call this method to get the converter for the
+    list items and then we call `TypeConverter.forType` to get a proper converter for lists.
+    */
+    val tpe = typeTag[U].tpe
+    (columnType, tpe) match {
+      case (argColumnType, TypeRef(_, Symbols.OptionSymbol, List(argScalaType))) =>
+        val argConverter = converter(argColumnType, argScalaType)
+        TypeConverter.forType[U](Seq(argConverter))
+      case (udt @ UserDefinedType(_, columns), _) =>
+        implicit val cm: ColumnMapper[U] = columnMapper[U]
+        new GettableDataToMappedTypeConverter[U](udt, udt.columnRefs)
+      case (ListType(argColumnType), TypeRef(_, _, List(argScalaType))) =>
+        val argConverter = converter(argColumnType, argScalaType)
+        TypeConverter.forType[U](Seq(argConverter))
+      case (SetType(argColumnType), TypeRef(_, _, List(argScalaType))) =>
+        val argConverter = converter(argColumnType, argScalaType)
+        TypeConverter.forType[U](Seq(argConverter))
+      case (MapType(keyColumnType, valueColumnType), TypeRef(_, _, List(keyScalaType, valueScalaType))) =>
+        val keyConverter = converter(keyColumnType, keyScalaType)
+        val valueConverter = converter(valueColumnType, valueScalaType)
+        TypeConverter.forType[U](Seq(keyConverter, valueConverter))
+      case (_, _) =>
+        TypeConverter.forType[U]
+    }
+  }
+
+  /** Returns a converter that can convert
+    * a Cassandra value of given `ColumnType` to Scala type given by `Type` */
+  private def converter(columnType: ColumnType[_], tpe: Type): TypeConverter[_] = {
+    type U = u forSome { type u }
+    implicit val tt = ReflectionUtil.typeToTypeTag[U](tpe)
+    converter[U](columnType)
+  }
+
+  /** Reads the column name of the column pointed by given column reference. */
+  private def columnName(columnRef: ColumnRef): String = {
```

Comment:
Good catch. It was more complex at some time before (when I was working on this ticket), because it required pattern-matching. Now it doesn't, so I'll inline it.